File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head.next:
            return True
        
        linkedstack = []
        curr = head
        while curr:
            linkedstack.append(curr.val)
            curr = curr.next

        while linkedstack:
            v = linkedstack.pop()
            if head.val == v:
                head = head.next
            else:
                return False

        return True

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    if (head==NULL || head->next==NULL) return head;

    struct ListNode* prev = NULL;
    struct ListNode* curr = head;
    while (curr) {
        struct ListNode* temp = curr->next;
        curr->next = prev;
        prev = curr;
        curr = temp;
    }

    return prev;
}

bool isPalindrome(struct ListNode* head) {
    if (head->next == NULL) return true;

    struct ListNode* reversed_head = reverseList(head);

    while (head && reversed_head) {
        if (head->val != reversed_head->val) {
            return false;
        }
        head = head->next;
        reversed_head = reversed_head->next;
    }

    return true;
}